src/azul/service/responseobjects/hca_response_v5.py

```diff
@@ -238,15 +242,58 @@ def validate(s: str) -> str:
             column_value = self.column_joiner.join(sorted(set(column_value)))
             row[column_name] = column_value
 
+    def _get_header(self, object_key):
+        try:
+            header = self.storage_service.head(object_key=object_key)
+        except self.storage_service.client.exceptions.ClientError as e:
+            if int(e.response['Error']['Code']) == 404:
+                return None
+            else:
+                raise e
+        else:
+            return header
+
+    def _can_use_cached_manifest(self, object_key):
+        """
+        Check if a s3 object was previously created and still exists in the bucket and won't be expiring soon.
+        :param object_key: S3 object key (eg. 'manifests/e0fabf97-7abb-5111-af97-810f1e736c71.tsv'
+        :return: boolean
+        """
+        header = self._get_header(object_key)
+        if header:
+            last_modified = header['LastModified']
+            assert isinstance(last_modified, datetime)
+            date_time_now = datetime.now(timezone.utc)
+            object_age = date_time_now - last_modified
+            seconds_until_expire = config.s3_object_lifetime_seconds - object_age.total_seconds()
+            if 'x-amz-expiration' in header:
+                # Verify x-amz-expiration matches the expiration date/time we expect
+                # example value: 'expiry-date="Fri, 21 Dec 2012 00:00:00 GMT", rule-id="Rule for testfile.txt"'
+                x_amz_expiration = parse_dict_header(header['x-amz-expiration'])
+                expiry_date = parse_date(x_amz_expiration['expiry-date'])
+                expected_expiry_date = date_time_now + timedelta(seconds=seconds_until_expire)
+                date_diff = expiry_date - expected_expiry_date
+                assert(abs(date_diff.total_seconds()) < config.s3_object_min_allowed_life_remaining)
```

Comment:
I think we want to use a different epsilon here. The purpose of s3_object_min_allowed_life_remaining is to ensure that the replacement is ready before the replaced object is deleted by the lifecycle rule. The epsilon for this comparison is not related to that concern, AFAICT. We should account for clock skew and maybe leap seconds, but that's it. Reading, https://aws.amazon.com/blogs/aws/keeping-time-with-amazon-time-sync-service/ I think 10s are a good epsilon for this.

Reply:
I'm a bit scared to make this an assert(). Note that typically there is a delay between `make deploy` and `make terraform` and during that time window the S3 expiration could differ dramatically from what the code expects it to be. I think we should change this to `log.error`.

Reply:
Question: if the expiration date from the `x-amz-expiration` header and the calculated value don't match, should we use the header value as the truth?

Reply:
Good point. Maybe we should. `x-amz-expiration` is the truth and that's the earliest time, the object could disappear. If the goal is to prevent a gap, we should use `x-amz-expiration`.